### ISSUE TITLE
Ablation study, reproducibility and determinism

### ISSUE THREAD

**Issue by Luugaaa** [ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star#23]
## Context

The current model with the following configuration produces very promising results. 
Overall Configuration (not exhaustive) : 
* region based training with minimal number of classes
* multichannel with phase + mag 
* initial LR = 0.001
* LR scheduler = Cosine 
* Optimizer = AdamW 
* Mag preprocessing : light clahe
* Phase preprocessing : heavy contrast enhancement  
* Augmentation : light spatial augmentation 

Most of those hyperparameters were chosen based on intuition and personal experience. We'd like to make ablation studies to understand which hyperparameters really improve the results. We would also like to investigate new possible source of improvement such as (not exhaustive) : 
* multichannel with mag & phase **and** with adjacent slices 
* stronger augmentations
* stronger/weaker/no preprocessing

## Goal 

To make it possible to draw conclusions from experiments with different hyperparameters, we would like to make the results as reproducible as possible. The best way to do that is to enable deterministic training. 

## Issue 

nnUnet does not currently support deterministic training, see [issue 1423 of the nnunet repo](https://github.com/MIC-DKFZ/nnUNet/issues/1423). And no evolution is planned regarding that matter. 

## Current investigation

I added the basics that should make the training deterministic, but it's not yet. 
```python 
seed=42
random.seed(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
np.random.seed(seed)
torch.manual_seed(seed)
# I'm on mps so no CUDA-specific seeding is needed
torch.use_deterministic_algorithms(True)
```

I then noticed that even the dataloaders were not deterministic, so changed [this code](https://github.com/MIC-DKFZ/nnUNet/blob/f1851fbaf2c53dcb51b079b60a01de528a7d0c17/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L675), to this : 

```python 
        if allowed_num_processes == 0:
            mt_gen_train = SingleThreadedAugmenter(dl_tr, None)
            mt_gen_val = SingleThreadedAugmenter(dl_val, None)
        else:
            train_seeds = [MASTER_SEED + i for i in range(allowed_num_processes)]
            
            num_val_processes = max(1, allowed_num_processes // 2)
            val_seeds = [MASTER_SEED + 1000 + i for i in range(num_val_processes)] # Use an offset to avoid overlap

            mt_gen_train = MultiThreadedAugmenter(data_loader=dl_tr, transform=None,
                                                num_processes=allowed_num_processes,
                                                num_cached_per_queue=max(6, allowed_num_processes // 2), 
                                                seeds=train_seeds,
                                                pin_memory=self.device.type == 'cuda', wait_time=0.002)
            
            mt_gen_val = MultiThreadedAugmenter(data_loader=dl_val, transform=None, 
                                                num_processes=num_val_processes,
                                                num_cached_per_queue=max(3, allowed_num_processes // 4),
                                                seeds=val_seeds,
                                                pin_memory=self.device.type == 'cuda',
                                                wait_time=0.002)
```

Now the dataloaders are deterministic, but the training still isn't. 

## Next step

* Remove augmentation, this may be the source of non determinism 

## Possible workaround 

I'm currently training 10 models (6/10 right now) on 40 epochs to have an estimate of the variability of the loss and metrics during training. If we cannot make nnunet deterministic, we can use this to have a better understanding of the impact of each hyperparameters. _Note : this will involves training (on 40 epochs) several times for each hyperparameters modification, this is very time consuming and not ideal_

**Comment by Luugaaa** [comment]
The random data augmentations (custom or default) are indeed the source of the non determinism ! When there are removed we have the following results, two trainings' (green & purple) metrics/loss overlap perfectly, the training is now deterministic !

<img width="941" height="301" alt="Image" src="https://github.com/user-attachments/assets/d375d424-e4bb-4ab4-bfc4-39835dba16d5" />

**Comment by Luugaaa** [comment]
Even when we override the augmentation code in the custom trainer (where we seed everything), the augmentation still breaks the determinism. 

### Investigation of the determinism of the default augmentations
The elastic deform from the [Spatial augmentation ](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/spatial/spatial.py#L19) is a source of determinism breaking. The source of the non determinism is [this code](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/spatial/spatial.py#L118-L139). The issue is the mixing of randomness from numpy and torch. The author does useless back and forth between torch and numpy, most likely because of how this method was originally implemented. It can be fixed (if you're willing to make changes in this module) by replacing it with this code : 
```python 
            offsets_np = np.random.randn(dim, *self.patch_size).astype(np.float32)

            # all the additional time elastic deform takes is spent here
            for d in range(dim):
                # Perform FFT, filtering, and IFFT entirely in NumPy
                tmp = np.fft.fftn(offsets_np[d])
                tmp = fourier_gaussian(tmp, sigmas[d])
                offsets_np[d] = np.fft.ifftn(tmp).real

                mx = np.max(np.abs(offsets_np[d]))
                offsets_np[d] /= (mx / np.clip(magnitude[d], a_min=1e-8, a_max=np.inf))
            
            # Convert the final NumPy array to a PyTorch tensor
            spatial_dims = tuple(list(range(1, dim + 1)))
            offsets = torch.from_numpy(offsets_np)
            offsets = torch.permute(offsets, (*spatial_dims, 0))
```

**Comment by Luugaaa** [comment]
It seems to be a common error made again and again. 

In the random transform (that encapsulates most augmentations), replace [this line](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/utils/random.py#L16) with this code : 
```python 
return {"apply_transform": np.random.rand() < self.apply_probability}
```

For the gaussian noise, replace [this line ](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/intensity/gaussian_noise.py#L22) with this code : 
```python
dct['apply_to_channel'] = np.random.rand(shape[0]) < self.p_per_channel
```
and [those lines](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/intensity/gaussian_noise.py#L37-L48) with this code : 
```python 
        if not isinstance(params['sigmas'], list):
            num_channels = sum(params['apply_to_channel'])
            # Generate noise with NumPy
            noise_np = np.random.normal(0, params['sigmas'], size=(1, *img_shape[1:]))
            # Convert to torch tensor
            gaussian = torch.from_numpy(noise_np.astype(np.float32))
            gaussian = gaussian.expand((num_channels, *[-1]*(len(img_shape) - 1)))
        else:
            # Generate noise with NumPy
            noise_np_list = [
                np.random.normal(0, i, size=(1, *img_shape[1:])) for i in params['sigmas']
            ]
            # Convert to torch tensors and concatenate
            gaussian = torch.cat(
                [torch.from_numpy(n.astype(np.float32)) for n in noise_np_list],
                dim=0
            )
        return gaussian
```

**Comment by Luugaaa** [comment]
For the gaussian blur, replace[ this line](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/noise/gaussian_blur.py#L102) with this : 
```python 
dct['apply_to_channel'] = np.random.rand(shape[0]) < self.p_per_channel
```

**And make sure to use benchmark=False !**

For the MultiplicativeBrightnessTransform replace [this line](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/intensity/brightness.py#L17) with this : 
```python
        apply_to_channel_np = np.where(np.random.rand(shape[0]) < self.p_per_channel)[0]
        apply_to_channel = torch.from_numpy(apply_to_channel_np)
```

The BrightnessAdditiveTransform is already good.

In the ContrastTransform your need to change [this line](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/intensity/contrast.py#L39) to : 
```python 
    apply_to_channel_np = np.where(np.random.rand(shape[0]) < self.p_per_channel)[0]
    apply_to_channel = torch.from_numpy(apply_to_channel_np)
```

In SimulateLowResolutionTransform, import numpy as np and change [this](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/spatial/low_resolution.py#L34-L37) into : 
```python 
    if self.allowed_channels is None:
        apply_to_channel_np = np.where(np.random.rand(shape[0]) < self.p_per_channel)[0]
        apply_to_channel = torch.from_numpy(apply_to_channel_np)
    else:
        apply_to_channel = [i for i in self.allowed_channels if np.random.rand() < self.p_per_channel]
```

Now in GammaTransform,  import numpy as np and change [this](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/intensity/gamma.py#L21-L23) into that : 
```python
    apply_to_channel_np = np.where(np.random.rand(shape[0]) < self.p_per_channel)[0]
    apply_to_channel = torch.from_numpy(apply_to_channel_np)
    
    retain_stats = torch.from_numpy(np.random.rand(len(apply_to_channel)) < self.p_retain_stats)
    invert_image = torch.from_numpy(np.random.rand(len(apply_to_channel)) < self.p_invert_image)
```

And finally in MirrorTransform, import numpy as np and change [this](https://github.com/MIC-DKFZ/batchgeneratorsv2/blob/07541d7eb5a4839aa4a5e494a123f3fe69ccfd4f/batchgeneratorsv2/transforms/spatial/mirroring.py#L14) into that : 
```python 
axes = [i for i in self.allowed_axes if np.random.rand() < 0.5]
```

**Comment by Luugaaa** [comment]
Now you've got yourself a nnunet training fully deterministic with the default augmentation pipeline !! 

<img width="941" height="301" alt="Image" src="https://github.com/user-attachments/assets/6296eb77-488f-4bfd-acde-0c1ec9c5488b" />

You should, of course, check if similar changes are needed when using other transforms.

**Comment by Luugaaa** [comment]
@jcohenadad I spent the day on this nnunet deterministic investigation, I'm thinking this might be interesting for SCT, as reproducibility was mentioned (I think) in a previous meeting. 

This is also a nnunet issue that has no answer, so I might as well give one to the people who were asking, and propose a PR both in nnunet and in [batchgeneratorsv2](https://github.com/MIC-DKFZ/batchgeneratorsv2) (same author, big source of non determinism) if you think it's worth it !

**Comment by jcohenadad** [comment]
> @jcohenadad I spent the day on this nnunet deterministic investigation, I'm thinking this might be interesting for SCT, as reproducibility was mentioned (I think) in a previous meeting.

Very cool investigation! And yes, it might be useful for the team. I suggest you mention it during our ivadomed and SCT meetings. 

> This is also a nnunet issue that has no answer, so I might as well give one to the people who were asking, and propose a PR both in nnunet and in [batchgeneratorsv2](https://github.com/MIC-DKFZ/batchgeneratorsv2) (same author, big source of non determinism) if you think it's worth it !

Yes! Definitely

**Comment by Luugaaa** [comment]
PRs submitted : [here](https://github.com/MIC-DKFZ/nnUNet/pull/2871) and [here](https://github.com/MIC-DKFZ/batchgeneratorsv2/pull/13). We'll see what they do with it !
